File: Hot 100/42. 接雨水.go
func trap(height []int) (ans int) {
    left, right := 0, len(height) - 1
    leftMax, rightMax := 0, 0
    for left < right {
        leftMax = max(leftMax, height[left])
        rightMax = max(rightMax, height[right])
        if leftMax < rightMax {
            ans += leftMax - height[left]
            left+=1
        } else {
            ans += rightMax - height[right]
            right-=1
        }
    }
    return
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: 刷题/Hot 100/42. 接雨水.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0, right = height.size()-1;
        int leftMax = 0;
        int rightMax = 0;
        int ans = 0;
        while(left < right) {
            leftMax = max(leftMax, height[left]);
            rightMax = max(rightMax, height[right]);
            if(leftMax < rightMax) {
                ans += leftMax - height[left];
                left++;
            } else {
                ans += rightMax - height[right];
                right--;
            }
        }
        return ans;
    }
};